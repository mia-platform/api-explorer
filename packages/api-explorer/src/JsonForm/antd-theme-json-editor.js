/* eslint-disable consistent-return */
/* eslint-disable no-underscore-dangle */
const JSONEditor = require('@json-editor/json-editor')

const antdTheme = JSONEditor.defaults.themes.bootstrap4.extend({
  getFormControl (label, input, description) {
    const labelText = label ? label.textContent : "";
    const descriptionText = description ? description.textContent : "";
    const group = document.createElement('div')

    if (label && (input.type === 'checkbox' || input.type === 'radio')) {
      group.classList.add('form-check')
      label.classList.add('form-check-label')
      input.classList.add('form-check-input')
      label.insertBefore(input, label.firstChild)
      group.appendChild(label)
    } else {
      group.classList.add('form-group')
      if (label) {
        label.classList.add('form-control-label')
        group.appendChild(label)
      }
      group.appendChild(input)
    }

    // Input Text.
    if (!input.type || input.type === 'text') { 
      input.classList.add('input-field', 'ant-input') 
    }

    if (description && labelText !== descriptionText) group.appendChild(description)

    return group
  },
  getButton (text, icon, title) {
    const el = this._super(text, icon, title)
    el.classList.add('ant-btn', 'ant-btn-primary')
    return el
  },
  getModal () {
    const el = this._super()
    el.style.background = "#fff"
    el.style.borderRadius = '4px'
    el.style.border = "none"
    el.style.marginTop = "5px"
    el.style.boxShadow = '0 2px 8px rgba(0, 0, 0, 0.15)'
    el.style.padding = '5px'
    el.style.position = 'absolute'
    el.style.zIndex = '10'
    el.style.display = 'none'
    el.style.fontSize = "14px"
    return el
  },
  getHeader (text) {
    const el = document.createElement('div')
    el.style.margin = '10px 0px 5px 0px'
    if (typeof text === 'string') {
      el.textContent = text
    } else {
      el.appendChild(text)
    }

    return el
  },
  getButtonHolder () {
    const el = document.createElement('div')
    el.classList.add('btt-holder-div')
    el.style.display = 'inline-flex'
    return el
  },
})

module.exports = antdTheme