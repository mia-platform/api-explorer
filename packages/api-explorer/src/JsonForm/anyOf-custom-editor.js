/* eslint-disable no-use-before-define */
/* eslint-disable no-underscore-dangle */
const baseCustomEditor = require('./get-custom-editor')

module.exports = () => baseCustomEditor('multiple').extend({
  build() {
    const self = this

    const switcher = this.theme.getSwitcher(self.display_text)
    switcher.multiple = true
    // const editors = self.display_text.map((text, i) => self.buildChildEditor(i))
    switcher.addEventListener('change', e => {
      e.preventDefault();
      e.stopPropagation();
      const selectedValues = Array.from(e.currentTarget.selectedOptions).map(selectedValue => selectedValue.value)
      if (selectedValues.length === 1) {
        self.errorMessageHtmlNode.style.display = 'none'
        self.editor_holder.style.display = 'block'
        self.switchEditor(self.display_text.indexOf(selectedValues[0]));
      } else {
        self.updateEditor(selectedValues)
      }
      self.onChange(true);
    });

    const response = this._super()
    this.container.removeChild(this.switcher)
    this.container.removeChild(this.editor_holder)
    this.switcher = switcher
    this.container.appendChild(switcher)
    this.container.appendChild(this.editor_holder)

    this.errorMessageHtmlNode = document.createElement('p')
    this.errorMessageHtmlNode.innerText = 'The selected schemas are not compatible with each other!'
    this.errorMessageHtmlNode.style.display = 'none'
    this.container.appendChild(this.errorMessageHtmlNode)

    return response
  },
  updateEditor(selectedValues) {
    const joinedValues = selectedValues.join('_')
    const mergedEditorIndex = editorIndexesMap.get(joinedValues)
    if(mergedEditorIndex) {
      this.switchEditor(mergedEditorIndex);
      return
    }
    const schemasIndexes = selectedValues.map(value => this.display_text.indexOf(value))
    const schemas = this.types.filter((_, index) => schemasIndexes.includes(index))
    const mergedSchemas = mergeSchemas(schemas)
    if (!mergedSchemas) {
      return
    }
    const index = this.types.length
    editorIndexesMap.set(joinedValues, index)
    this.types.push(mergedSchemas)
    this.editors.push(false)
    this.switchEditor(index)
  }
})

const editorIndexesMap = new Map()

const mergeSchemas = (schemas) => {
  const schemasType = schemas.reduce((type, currentSchema) => {
    if (type === 'any') {
      return currentSchema.type
    }
    return currentSchema.type === type
      ? type
      : undefined
  }, 'any')
  if (!schemasType) {
    this.editor_holder.style.display = 'none'
    this.errorMessageHtmlNode.style.display = 'block'
    return undefined
  }
  const merger = mergersMap[schemasType]
  const mergedSchemas = merger
    ? merger(schemas, schemasType)
    : mergersMap.default(schemas, schemasType)
  return mergedSchemas
}

const mergeArrays = schemas => {
  const schemasItems = schemas.map(schema => schema.items)
  const mergedItems = mergeSchemas(schemasItems)
  if (!mergedItems) {
    return undefined
  }
  return { type: 'array', items: mergedItems }
}
const mergeObjects = schemas => {
  const mergedProperties = schemas.reduce((propsAccumulator, currentSchema) => {
    return { ...propsAccumulator, ...currentSchema.properties }
  }, {})

  return {
    type: 'object',
    properties: mergedProperties,
  }
}

const mergersMap = {
  array: mergeArrays,
  object: mergeObjects,
  default: (_, schemasType) => { return { type: schemasType } }
}
