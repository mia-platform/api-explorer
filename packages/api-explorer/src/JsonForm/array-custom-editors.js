/* eslint-disable no-underscore-dangle */
const baseArrayCustomEditor = require('./get-custom-editor')

function setControlsStyle(controls) {
  if(controls) {
      
    controls.style.display = 'grid'
    controls.style.gridTemplateColumns = 'repeat(3, min-content)'
    controls.style.gridGap = '10px'
  }
}
module.exports = () => baseArrayCustomEditor('array').extend({
  addControls(){
    const response = this._super()
    setControlsStyle(this.controls)
    return response
  },
  refreshValue(force) {
    const response = this._super(force)
    if (this.controls) {
      const currentStyle = this.controls.style.display
      if (currentStyle && currentStyle !== 'none'){
        setControlsStyle(this.controls)
      }
    }
    return response
  }
})