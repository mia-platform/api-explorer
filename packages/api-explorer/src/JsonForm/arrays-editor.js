module.exports = (classReference) => class arrayCustomEditor extends classReference {

  postBuild() {
    super.postBuild()

    if (this.parent && this.header) {
      this.header.style.textTransform = 'unset'
      this.header.style.fontSize = '14px'
      this.header.style.fontWeight = '400'

      if (
        this.jsoneditor &&
        this.jsoneditor.schema &&
        Array.isArray(this.jsoneditor.schema.required) &&
        this.jsoneditor.schema.required.includes(this.key) &&
      ) {
        this.header.classList.add('required')
      }
    }
  }
}
