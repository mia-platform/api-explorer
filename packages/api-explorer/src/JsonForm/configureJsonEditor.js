import it from '../../i18n/it.json'

import antdTheme from "./antd-theme-json-editor";
import notCustomEditor from "./not-custom-editor";
import anyOfEditor from "./anyOf-custom-editor";
import objectsEditor from "./objects-editor";

function setDefaultCustomization (JSONEditor) {
  const editorsKeys = Object.keys(JSONEditor.defaults.editors)

  editorsKeys.forEach(key => {
    JSONEditor.defaults.editors[key] = class Customization extends JSONEditor.defaults.editors[key] {
      postBuild() {
        super.postBuild()

        if (this.schema && this.schema.type && this.label) {
          this.label.innerHTML = `${this.label.innerHTML} <em style="opacity: 0.5">(${this.schema.type})</em>`
        }

        if (this.editjson_control) {
          this.editjson_control.classList.add('ant-btn-sm')
          this.editjson_control.style.margin = '0px 8px'
          this.editjson_control.style.borderRadius = '4px'
        }

        if (this.addproperty_button) {
          this.addproperty_button.classList.add('ant-btn-sm')
          this.addproperty_button.style.margin = '0px 8px'
          this.addproperty_button.style.borderRadius = '4px'
        }

        if (this.collapse_control) {
          this.collapse_control.classList.add('ant-btn-sm')
        }

        if (this.toggle_button) {
          this.toggle_button.classList.add('ant-btn-sm')
        }

        if (this.container !== undefined && this.title !== undefined) {
          const classContainer = (key === 'upload' || key === 'base64') ? 'mia-container-upload-file-wrapper' : 'mia-container-wrapper'
          this.container.classList.add(classContainer)
        }
      }
    }
  });
}
module.exports = function configureJSONEditor(JSONEditor, intl, setFormSubmissionListener) {
  if (JSONEditor.hasBeenEdited) return
  JSONEditor.hasBeenEdited = true
  JSONEditor.defaults.languages.it = it
  JSONEditor.defaults.language = intl.locale

  JSONEditor.defaults.themes.antdTheme = antdTheme(JSONEditor)

  setDefaultCustomization(JSONEditor)

  JSONEditor.defaults.editors.not = notCustomEditor(JSONEditor.defaults.editors.multiple)
  JSONEditor.defaults.editors.anyOf = anyOfEditor(intl, setFormSubmissionListener, JSONEditor.defaults.editors.multiple)
  JSONEditor.defaults.editors.object = objectsEditor(JSONEditor.defaults.editors.object)
  // eslint-disable-next-line consistent-return
  JSONEditor.defaults.resolvers.unshift((scheme) => {
    // If the schema can be of any type
    if (
      (scheme.type === "string" &&
        scheme.media &&
        scheme.media.binaryEncoding === "base64") ||
      scheme.format === "binary"
    ) {
      return "base64";
    }
  });

  // eslint-disable-next-line consistent-return
  JSONEditor.defaults.resolvers.unshift((scheme) => {
    if (scheme.not && scheme.not.type) {
      return 'not'
    }
  })

  // eslint-disable-next-line consistent-return
  JSONEditor.defaults.resolvers.unshift((scheme) => {
    if (scheme.anyOf) {
      return 'anyOf'
    }
  })
}
