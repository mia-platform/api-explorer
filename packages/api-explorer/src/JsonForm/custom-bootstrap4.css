/** json-editor */
.card .row .form-group .form-text {
    text-align: right;
    font-size: 0.9em;
    opacity: 0.8;
}

.btn-holder-div > button {
    margin: 5px;
}

.mia-container-wrapper {
    display: grid;
    grid-template-columns: 1fr auto;
    grid-template-rows: 40px 1fr;
    grid-template-areas: 'card-title btn-group' 'card-body card-body';
    align-items: center;
}

.mia-container-wrapper .card-title {
    grid-area: card-title;
    display: flex !important;
    margin: 0;
    min-width: 0;
}

.mia-container-wrapper .card-title .ant-btn-sm {
    grid-area: card-title;
    display: flex;
    align-items: center;
    justify-content: center;
    color: rgba(0, 0, 0, 0.65);
    margin: 0 !important;
    border: 0;
    width: 24px;
    height: 24px;
    background-color: transparent;
}

.mia-container-wrapper .btn-group {
    grid-area: btn-group;
    margin: 0 !important;
}

.mia-container-wrapper .btn-group textarea {
    margin: 0 !important;
    resize: both;
}
.mia-container-wrapper .btn-group .json-editor-btntype-copy,
.mia-container-wrapper .btn-group .json-editor-btntype-cancel,
.mia-container-wrapper .btn-group .json-editor-btntype-save {
    margin: 4px 4px 0 0
}

.mia-container-wrapper .card-title label {
    font-size: 18px;
    line-height: 25px;
    color: #666666;
    text-transform: uppercase;
    font-weight: bold;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    margin: 0
}

.mia-container-wrapper .card-body {
    grid-area: card-body;
}

.form-group {
    display: grid;
    grid-template-columns: 200px 1fr;
    align-items: center;
    margin: 0px;
    padding: 10px 0px;
}

.form-group label {
    margin: 0px;
}

.btn:focus, .btn-secondary:focus {
    box-shadow: none !important;
}

/** MIA CSS */

.mia-container-wrapper {
    border-radius: 4px;
    border: solid 1px #d8d8d8;
    background: #fafafa;
    padding: 5px
}

.mia-container-wrapper.card {
    display: block;
}

.mia-container-wrapper .btn-group {
    width: max-content;
}
/** antd */

.ant-btn svg {
    margin-right: 5px;
}

/** general */

* {
    outline: none;
}
