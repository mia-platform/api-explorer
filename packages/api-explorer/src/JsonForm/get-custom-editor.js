/* eslint-disable no-underscore-dangle */

const getCustomEditor = (JSONEditor, key) =>{
  // const extendsEditorKey = JSONEditor.defaults.editors[key]
  // console.log('extendsEditorKey', extendsEditorKey)
  // extendsEditorKey.build = () => {
  //   // key (the schema typology) is set to multiple when it's
  //   // anyOf or similar, jsoneditor uses multiple for some of
  //   // these kind of schemas, in case the current key is multiple
  //   // during schema change we want the editor to be cleared out.
  //   if (key === 'multiple') {
  //     this.keep_values = false
  //   }
  //   const buildResponse = this._super()
  //   if(this.addproperty_add) {
  //     this.addproperty_add.style.marginLeft = '10px'
  //   }
  //   return buildResponse
  // }
  // extendsEditorKey.setContainer = (container) => {
  //   this.container = container;
  //   this.container.style.paddingTop = "5px";
  //   this._super(this.container)
  // }
  // return extendsEditorKey
  return JSONEditor.defaults.editors[key].extends({
    // setContainer (container) {
    //   this.container = container;
    //   this.container.style.paddingTop = "5px";
    //   this._super(this.container)
    // },
    build() {
      // key (the schema typology) is set to multiple when it's
      // anyOf or similar, jsoneditor uses multiple for some of
      // these kind of schemas, in case the current key is multiple
      // during schema change we want the editor to be cleared out.
      if (key === 'multiple') {
        this.keep_values = false
      }
      const buildResponse = this._super()
      if(this.addproperty_add) {
        this.addproperty_add.style.marginLeft = '10px'
      }
      return buildResponse
    }
  });
}
module.exports = getCustomEditor;
