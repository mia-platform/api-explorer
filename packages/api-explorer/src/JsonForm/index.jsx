import '@fortawesome/fontawesome-free/js/fontawesome'
import '@fortawesome/fontawesome-free/js/solid'
import '@fortawesome/fontawesome-free/js/regular'
import '@fortawesome/fontawesome-free/js/brands'

import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {injectIntl} from 'react-intl'
import {JSONEditor} from '@json-editor/json-editor'
import {Spin, Alert} from 'antd'

import configureJsonEditor from './configureJsonEditor'
import resolveReference from './resolveReference'

import './bootstrap4.css'
import './custom-bootstrap4.css'

class JsonForm extends Component {
  constructor(props) {
    super(props);
    this.editor = null;
    this.ref = null;

    this.state = {
      error: null,
      jsonSchema: null
    }
    this.unmounted = false
  }

  componentDidMount() {
    const {schema} = this.props
    resolveReference(schema)
      .then(convertedSchema => {
        if (!this.unmounted) {
          return this.setState({
            jsonSchema: convertedSchema
          })
        }
        return null
      }).catch(err => {
        if (err && !this.unmounted) {
          return this.setState({error: err.message})
        }
        return null
      })
  }

  componentWillUnmount() {
    this.unmounted = true
  }

  createEditor(element, jsonSchema) {
    const {intl, onChange, setFormSubmissionListener, title} = this.props
    if (this.editor === null) {
      configureJsonEditor(JSONEditor, intl, setFormSubmissionListener)
      this.editor = new JSONEditor(element, {
        schema: {
          ...jsonSchema,
          title
        },
        show_opt_in: false,
        prompt_before_delete: false,
        form_name_root:"",
        iconlib: 'fontawesome5',
        theme: 'antdTheme'
      });
      this.editor.on('change', () => onChange(this.editor.getValue()))
    }
  }

  render() {
    const {onSubmit} = this.props
    const {error, jsonSchema} = this.state
    if (error) {
      return (
        <Alert
          message={error}
          type={error}
          showIcon
        />
      )
    }
    if (jsonSchema) {
      return (
        <form
          ref={r => this.createEditor(r, jsonSchema)}
          onSubmit={(e) => {
            e.preventDefault()
            onSubmit()
          }}
        >
          <button type="submit" style={{ display: "none" }} />
        </form>
      );
    }
    return <Spin />
  }
}

JsonForm.propTypes = {
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  schema: PropTypes.object.isRequired,
  setFormSubmissionListener: PropTypes.func.isRequired,
  intl: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired
}
export default injectIntl(JsonForm)
