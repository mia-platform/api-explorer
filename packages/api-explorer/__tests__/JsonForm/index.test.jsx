import React from 'react'
import { mountWithIntl } from 'enzyme-react-intl'

import JsonForm from '../../src/JsonForm'
import SCHEMA_WITH_REF_ROOT_AND_NESTED from '../testdata/config-root-ref-and-nested.json'
import EXPECTED_WITH_REF_ROOT_AND_NESTED from '../testdata/config-root-ref-and-nested.expected.json'
import JSON_EDITOR_MAX_STACK_SCHEME_BUG from '../testdata/json-editor-bug-max-stack.json'

describe('JSONForm ', () => {
  const props = {
    schema: {
      type: 'object',
      properties: {
        petId: {
          type: 'integer',
          description: 'ID of pet to return',
          format: 'int64'
        },
        petPedigree: {
          type: 'object',
          properties: {
            petName: {
              type: 'string',
              description: 'Name of pet to return'
            }
          }
        }
      },
      required: ['petId', 'petPedigree']
    },
    onChange: jest.fn(),
    onSubmit: jest.fn(),
    setFormSubmissionListener: jest.fn(),
    title: 'The Title'
  }

  it('snapshot', async () => {
    /**
    * mock because of random id on input/label
    * https://github.com/mia-platform/json-editor/blob/master/src/themes/bootstrap4.js#L129
    */
   
    const originalDateNow = Date.now
    const originalRandom = Math.random
    const originalRequestAnimationFrame = window.requestAnimationFrame

    Date.now = jest.fn().mockReturnValue(1589886640576)
    Math.random = jest.fn().mockReturnValue(0.5)

    /**
    * window.requestAnimationFrame has been mocked to avoid a non-deterministic behaviour of the test.
    * Removing the line below could make the test produce a different snapshot from the expected one.
    */
    window.requestAnimationFrame = jest.fn()

    const element = await mountJsonForm(props)
    expect(element.getDOMNode()).toMatchSnapshot()

    Date.now = originalDateNow
    Math.random = originalRandom
    window.requestAnimationFrame = originalRequestAnimationFrame
  })

  it('when form submits, calls onSubmit prop', async () => {
    const element = await mountJsonForm(props)
    const mockEvent = { preventDefault: jest.fn() }
    element.find('form').prop('onSubmit')(mockEvent)
    expect(props.onSubmit).toHaveBeenCalledTimes(1)
    expect(mockEvent.preventDefault).toHaveBeenCalledTimes(1)
  })

  it('on parsing error, shows error', async () => {
    const element = await mountJsonForm({
      ...props,
      schema: {'$ref': '#/foo'}
    })
    expect(element.getDOMNode()).toMatchSnapshot()
  })

  it('converts schema correctly', async () => {
    const schemaWithRef =  {
      components: {
        address: {
          type: "object",
          properties: {
            street_address: { type: "string" },
            city: { type: "string" },
            state: { type: "string" }
          },
          required: ["street_address", "city", "state"]
        }
      },
      $ref: "#/components/address"
    };
    const element = await mountJsonForm({...props, schema: schemaWithRef})
    expect(element.find('JsonForm').instance().jsonSchema).toEqual({
      definitions: {
        components: {
          definitions: {
            address: {
              type: "object",
              properties: {
                street_address: { type: "string" },
                city: { type: "string" },
                state: { type: "string" }
              },
              required: ["street_address", "city", "state"]
            }
          }
        }
      },
      type: "object",
      properties: {
        street_address: { type: "string" },
        city: { type: "string" },
        state: { type: "string" }
      },
      required: ["street_address", "city", "state"]
    })
  })

  it('converts schema correctly with nested ref', async () => {
    const element = await mountJsonForm({...props, schema: SCHEMA_WITH_REF_ROOT_AND_NESTED})
    expect(element.find('JsonForm').instance().jsonSchema).toEqual(EXPECTED_WITH_REF_ROOT_AND_NESTED)
  })

  it('scheme with recursion in object properties (json-editor bug)', async () => {
    const element = await mountJsonForm({...props, schema: JSON_EDITOR_MAX_STACK_SCHEME_BUG})
    expect(element.find('form')).toHaveLength(1)
  })
})

function mountJsonForm (props) {
  return new Promise((resolve) => {
    const element = mountWithIntl(<JsonForm {...props} />)
    setTimeout(() => {
      element.update()
      resolve(element)
    }, 0)
  })
}
