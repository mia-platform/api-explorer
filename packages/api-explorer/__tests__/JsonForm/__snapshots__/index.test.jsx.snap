// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONForm  on parsing error, shows error 1`] = `
<span>
  Token "foo" does not exist.
</span>
`;

exports[`JSONForm  snapshot 1`] = `
<form
  data-theme="antdTheme"
>
  <button
    style="display: none;"
    type="submit"
  />
  <div
    class="mia-container-wrapper"
    data-schemaid="root"
    data-schemapath="root"
    data-schematype="object"
    style="position: relative;"
  >
    <h3
      class="card-title"
      style="display: inline-block;"
    >
      <button
        class="btn btn-secondary btn-sm ant-btn json-editor-btn-collapse json-editor-btntype-toggle ant-btn-sm"
        style="margin: 0px 10px 0px 0px; border: 0px; background: transparent;"
        title="Collapse"
        type="button"
      >
        <i
          class="fas fa-caret-down"
        />
        <span>
           
        </span>
      </button>
      <label>
        The Title
      </label>
    </h3>
    <span
      class="btn-group"
      style="display: inline-flex; margin: 5px 0px 0px 10px; float: right;"
    >
      <div
        style="box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15); position: absolute; z-index: 10; display: none; background: rgb(255, 255, 255); border-radius: 4px; margin-top: 5px; padding: 5px; font-size: 14px;"
      >
        <textarea
          class="form-control"
          style="height: 170px; width: 300px; display: block;"
        />
        <button
          class="btn btn-secondary btn-sm ant-btn json-editor-btn-save json-editor-btntype-save"
          title="Save"
          type="button"
        >
          <i
            class="fas fa-save"
          />
          <span>
             Save
          </span>
        </button>
        <button
          class="btn btn-secondary btn-sm ant-btn json-editor-btn-copy json-editor-btntype-copy"
          title="Copy"
          type="button"
        >
          <i
            class="fas fa-copy"
          />
          <span>
             Copy
          </span>
        </button>
        <button
          class="btn btn-secondary btn-sm ant-btn json-editor-btn-cancel json-editor-btntype-cancel"
          title="Cancel"
          type="button"
        >
          <i
            class="fas fa-ban"
          />
          <span>
             Cancel
          </span>
        </button>
      </div>
      <div
        style="box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15); position: absolute; z-index: 10; display: none; background: rgb(255, 255, 255); border-radius: 4px; margin-top: 5px; padding: 5px; font-size: 14px;"
      >
        <div
          class="property-selector"
          style="width: 295px; max-height: 160px; padding: 5px 0px 5px 5px; overflow-y: auto; overflow-x: hidden;"
        >
          <div
            class="form-group"
            style="margin-top: 0px; padding-top: 0px; margin-bottom: 0px; padding-bottom: 0px; height: auto;"
          >
            <div
              class="form-check"
            >
              <input
                class="form-check-input"
                disabled=""
                style="width: auto;"
                type="checkbox"
              />
              <label
                class="form-check-label"
              >
                Â petId
              </label>
            </div>
          </div>
        </div>
        <input
          class="form-control"
          placeholder="Property name..."
          style="width: 220px; margin-bottom: 0px; display: inline-block;"
          type="text"
        />
        <button
          class="btn btn-secondary btn-sm ant-btn json-editor-btn-add json-editor-btntype-add"
          title="add"
          type="button"
        >
          <i
            class="fas fa-plus"
          />
          <span>
             add
          </span>
        </button>
        <div
          style="clear: both;"
        />
      </div>
      <button
        class="btn btn-secondary btn-sm ant-btn json-editor-btn-edit json-editor-btntype-editjson ant-btn-sm"
        style="margin: 0px 8px; border-radius: 4px;"
        title="Edit JSON"
        type="button"
      >
        <i
          class="fas fa-pen"
        />
        <span>
           JSON
        </span>
      </button>
      <button
        class="btn btn-secondary btn-sm ant-btn json-editor-btn-edit_properties json-editor-btntype-properties ant-btn-sm"
        style="margin: 0px 8px; border-radius: 4px;"
        title="Object Properties"
        type="button"
      >
        <i
          class="fas fa-list"
        />
        <span>
           Properties
        </span>
      </button>
    </span>
    <div
      style="display: none;"
    />
    <div
      class="card card-body mb-3 bg-light"
    >
      <div>
        <div>
          <div
            class="row"
          >
            <div
              class="col-md-12"
              data-schemapath="root.petId"
              data-schematype="integer"
            >
              <div
                class="form-group"
              >
                <label
                  class="required"
                >
                  petId
                </label>
                <input
                  class="form-control"
                  data-schemaformat="int64"
                  name="root[petId]"
                  type="text"
                />
                <div />
                <small
                  class="form-text"
                >
                  ID of pet to return
                </small>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</form>
`;
